### PR TITLE
fix: support pnpm lockfile v6 use of yaml complex mapping keys

### LINKED ISSUE #1015: [Bug]: rules_js fails to parse pnpm lockfile v6 for packages with long lists of peer dependencies
### What happened?

Apparently pnpm lockfile v6 uses complex yml mapping keys for packages with long lists of peer dependencies which is not parsed correctly by rules_js. I get the below exception

```
Repository rule npm_translate_lock defined at:
  /private/var/tmp/_bazel_theys/c47bd7bac752f67cacbe3862d74d80c0/external/aspect_rules_js/npm/private/npm_translate_lock.bzl:101:37: in <toplevel>
ERROR: An error occurred during the fetch of repository 'npm':
   Traceback (most recent call last):
	File "/private/var/tmp/_bazel_theys/c47bd7bac752f67cacbe3862d74d80c0/external/aspect_rules_js/npm/private/npm_translate_lock.bzl", line 78, column 58, in _impl
		importers, packages = translate_to_transitive_closure(
	File "/private/var/tmp/_bazel_theys/c47bd7bac752f67cacbe3862d74d80c0/external/aspect_rules_js/npm/private/transitive_closure.bzl", line 127, column 53, in translate_to_transitive_closure
		package, package_info = _gather_package_info(package_path, package_snapshot)
	File "/private/var/tmp/_bazel_theys/c47bd7bac752f67cacbe3862d74d80c0/external/aspect_rules_js/npm/private/transitive_closure.bzl", line 82, column 17, in _gather_package_info
		fail(msg)
Error in fail: expected package resolution to have a name field
ERROR: /Users/theys/r9/R9/WORKSPACE:257:19: fetching npm_translate_lock rule //external:npm: Traceback (most recent call last):
	File "/private/var/tmp/_bazel_theys/c47bd7bac752f67cacbe3862d74d80c0/external/aspect_rules_js/npm/private/npm_translate_lock.bzl", line 78, column 58, in _impl
		importers, packages = translate_to_transitive_closure(
	File "/private/var/tmp/_bazel_theys/c47bd7bac752f67cacbe3862d74d80c0/external/aspect_rules_js/npm/private/transitive_closure.bzl", line 127, column 53, in translate_to_transitive_closure
		package, package_info = _gather_package_info(package_path, package_snapshot)
	File "/private/var/tmp/_bazel_theys/c47bd7bac752f67cacbe3862d74d80c0/external/aspect_rules_js/npm/private/transitive_closure.bzl", line 82, column 17, in _gather_package_info
		fail(msg)
```

This exception is thrown while interpreting the following entry in the pnpm v6 lockfile (note: I redacted the module names because these are internally published packages and I'd prefer not to share them). It appears that pnpm's yaml serializer uses complex mapping keys (https://yaml.org/spec/1.2.2/#example-mapping-between-sequences) for very long key names and the rules_js yaml parser does not interpret this correctly.

```
  ? /@myorg/foobar@1(@babel/runtime@7.20.6)(@myorg/pkg1@1)(@myorg/pkg2@1)(@myorg/pkg3@1)(@myorg/pkg4@1)(@myorg/pkg4@1)(@myorg/pkg5@1)(@myorg/pkg5@1)(@myorg/pkg6@1)(@myorg/pkg7@1)(@myorg/pkg8@1)(@myorg/pkg9@1)(@myorg/pkg10@1)(@myorg/pkg11@1)(@myorg/pkg12@1)(@myorg/pkg13@1)(@myorg/pkg14@1)(@myorg/pkg15@1)(@myorg/pkg16@1)(@myorg/pkg17@1)(@myorg/pkg18@1)(@myorg/pkg19@1)(@myorg/pkg20@1)(@myorg/pkg21@1)(@myorg/pkg22@1)(@myorg/pkg23@1)(@myorg/pkg24@1)(@myorg/pkg25@1)(@myorg/pkg26@1)(@myorg/pkg27@1)(@myorg/pkg28@1)(@myorg/pkg29@1)(@myorg/pkg30@1)(classnames@2.3.1)(react-dom@17.0.2)(react@17.0.2)
  : resolution: {integrity: ...}
    peerDependencies:
      '@babel/runtime': ^7.17.2
```

### Version

This issue appears in rules_js@v1.24.0 and all versions thereafter

### How to reproduce

```shell
In any lockfile, manually change an entry to use a complex yaml mapping key, then build any bazel target.
```

### Any other information?

_No response_

### PR DESCRIPTION
**Pull request by jbedard**
Fix #1015

### CODE REVIEW

**Comment by kormide** [comment]
Can we get some more test cases where complex maps are mixed with other structures?

**Comment by kormide** [comment]
Do complex maps only ever have two keys?

**Comment by kormide** [comment]
Can you add an example of how it's used in the pnpm lock test?

**Comment by kormide** [comment]
Complex maps support things like lists as keys, right? Can we get an example of that?

**Comment by jbedard** [comment]
> Can we get some more test cases where complex maps are mixed with other structures?

I'm adding more atm, there's definitely some cases not working atm.

> Can you add an example of how it's used in the pnpm lock test?

I've added a few.

> Complex maps support things like lists as keys, right? Can we get an example of that?

Yes it does, and I'll give it a shot, but we don't really care about that case so I might just leave it as unsupported if it requires any effort.

File: npm/private/test/yaml_tests.bzl [hunk 1]
```diff
@@ -292,6 +312,37 @@ baz: 5
...
+  : ? c
+    : - 1
+      - 2
+"""))
+
+    asserts.equals(env, {"a": {"b": {"c": 1, "d": 2}}}, parse("""\
+a:
+  ? b
+  : c: 1
+    d: 2
```
Comment (kormide) [inline npm/private/test/yaml_tests.bzl:341]: Can you add another test where you go shallower after going deep?

File: npm/private/yaml.bzl [hunk 2]
```diff
@@ -81,6 +81,29 @@ def _handle_CONSUME_SPACE(state, input, stack, starlark):
         # Consume any space following the {
         stack.append(_new_CONSUME_SPACE_FLOW())
 
+    elif input == "?":
+        stack.pop()
+
+        # We are at the beginning of a new complex map
```
Comment (kormide) [inline npm/private/yaml.bzl:87]: Is it more correct to say that we're at the beginning of a complex key, since a map can have complex and normal keys?

**Review by kormide — APPROVED** [review]

File: npm/private/yaml.bzl [hunk 3]
```diff
@@ -81,6 +81,36 @@ def _handle_CONSUME_SPACE(state, input, stack, starlark):
         # Consume any space following the {
         stack.append(_new_CONSUME_SPACE_FLOW())
 
+    elif input == "?":
```
Comment (jbedard) [inline npm/private/yaml.bzl:84]: So these key/value pairs are initiated different then the standard ":" style. With the ":" we might not know we are in a map until we reach the ":", and at that point we've already parsed the key. While the "?" comes before the key so we know up front we're in a map, but we  don't know the key yet.

**Comment by kormide** [comment]
Looking great so far. Can we add a few tests involving preceeding and trailing spaces around keys? Can spaces be part of a key?

**Review by kormide — APPROVED** [review]

File: npm/private/yaml.bzl (hunk 3 above)
Comment (kormide) [reply npm/private/yaml.bzl:84]: nit: add a comment here that the parser currently only supports string keys for complex maps

**Pull request merged**